### PR TITLE
Add cache property for applications to COUModel

### PR DESCRIPTION
**Pull request by rgildein**
Like this it will be possible to access all application in model, without calling get_applications. This will be used in planning stage, where we are not running async functions and for any app, which require using other applications. e.g. ceph-osd need to check nova-compute status.

With this feature, we can easily access all colocated nova-compute and ceph-mon during pre- and post-upgrade steps in ceph-osd.

### CODE REVIEW

File: cou/utils/juju_utils.py [hunk 1]
```diff
@@ -201,6 +201,22 @@ def __init__(self, name: Optional[str]):
...
+        """Return cached applications.
+
+        This property represents the cached output of the get_applications function.
+
+        :return: dictionary of COUApplication
+        :rtype: dict[str, COUApplication]
+        :raises ValueError: When get_applications has not yet been called.
+        """
+        if self._applications is None:
+            raise ValueError("The get_applications has not yet been called.")
```
Comment (gabrielcocenza) [inline cou/utils/juju_utils.py:217]: Why not call `get_applications` here instead of raising an error?

Comment (rgildein) [reply cou/utils/juju_utils.py:217]: async?

Comment (gabrielcocenza) [reply cou/utils/juju_utils.py:217]: Forgot about that :sweat_smile:

**Review by gabrielcocenza — APPROVED** [review]

**Comment by rgildein** [comment]
Please check #280

**Review by agileshaw — APPROVED** [review]
LGTM

**Pull request merged**